isomerpages/isomercms-backend
# Feat(token): Improve token selection logic

## Problem

The current token selection algorithm uses random selection, which causes the following problems:
- Difficulty in monitoring token usage as worst token might deviate from overall system usage
- All tokens have close hourly reset time, resulting in usage spike just before reset and dips right after reset
- Does not maximise overall token capacity
- Cannot be easily scaled upward during peak usage

Closes IS-173

## Solution

1. For active tokens pool, pick the token with the earliest reset time thats below threshold.
2. Switch to reserve tokens pool when all active tokens have reached threshold. Each instance would pick only 1 token until its exhausted.
3. Switch back to active token pool after an hour and all active tokens have been reset. 

More details presented [here](https://www.notion.so/opengov/Github-Tokens-Selection-RFC-IS173-af89ac51fb53455a82f7be4c18178483).

**Breaking Changes**

  - This PR requires is_reserved:boolean and reset_time:datetime columns to be added to access_token table in database
  - db migration script should be executed

**Improvements**:

- Allows for monitoring of system capacity through number of exhausted active and reserved tokens
- Optimises overall system token capacity
- Additional reserved tokens can be added during peak periods by insert into database

## Tests

Test for tokens are being inserted into requests and response correctly.
1. Ensure that there is a valid active token in access_tokens table.
2. Login via email user and make edits to `a-test-v4` page. 
3. Ensure that no errors occurred.

Unit tests are added to test TokenServices and its components. 

## Deploy Notes

**Database Schema Changes**
- Run database migration script.
